Kailyn Philpot
003071160

# Breaking down the 90s Hit Formula




The "Classic 90s Hits" dataset from Kaggle features a collection of popular songs from the 1990s, capturing key musical characteristics. It includes variables such as song title, artist, tempo, duration, and various audio features like danceability and energy.


In [31]:
import pandas as pd

# Load dataset
df = pd.read_csv('/content/1990sClassicHits.csv')
df.head()

,Track,Artist,Duration,Time_Signature,Danceability,Energy,Key,Loudness,Mode,Speechiness,Acousticness,Instrumentalness,Liveness,Valence,Tempo,Popularity,Year
0,Everybody Everybody,Black Box,4:11,4,0.645,0.795,7,-12.095,1,0.0527,0.3470,0.003890,0.1040,0.535,112.067,87,1990
1,Close To You,Maxi Priest,3:45,4,0.590,0.727,6,-7.955,1,0.0706,0.0259,0.000065,0.0827,0.356,121.966,85,1990
2,Do Me!,Bell Biv DeVoe,3:56,4,0.615,0.534,9,-6.719,0,0.0386,0.1060,0.000018,0.0607,0.193,135.917,84,1990
3,Everything,Jody Watley,4:05,4,0.704,0.225,6,-14.454,0,0.0994,0.9020,0.657000,0.1060,0.243,120.006,82,1990
4,What It Takes,Aerosmith,3:21,4,0.672,0.655,10,-5.021,0,0.0310,0.0362,0.000000,0.1170,0.555,134.948,82,1990


### SECTION I: DATASET OVERVIEW AND STRUCTURE
---

#### 1. Shape and Structure of the Dataset


In [16]:
# Shape of dataset
print("Dataset Shape:", df.shape)

# First few rows
print("\nFirst 5 rows:")
display(df.head())

# Last few rows
print("\nLast 5 rows:")
display(df.tail())

# General info
print("\nData Info:\n")
df.info()

Dataset Shape: (919, 17)

First 5 rows:


,Track,Artist,Duration,Time_Signature,Danceability,Energy,Key,Loudness,Mode,Speechiness,Acousticness,Instrumentalness,Liveness,Valence,Tempo,Popularity,Year
0,Everybody Everybody,Black Box,4:11,4,0.645,0.795,7,-12.095,1,0.0527,0.3470,0.003890,0.1040,0.535,112.067,87,1990
1,Close To You,Maxi Priest,3:45,4,0.590,0.727,6,-7.955,1,0.0706,0.0259,0.000065,0.0827,0.356,121.966,85,1990
2,Do Me!,Bell Biv DeVoe,3:56,4,0.615,0.534,9,-6.719,0,0.0386,0.1060,0.000018,0.0607,0.193,135.917,84,1990
3,Everything,Jody Watley,4:05,4,0.704,0.225,6,-14.454,0,0.0994,0.9020,0.657000,0.1060,0.243,120.006,82,1990
4,What It Takes,Aerosmith,3:21,4,0.672,0.655,10,-5.021,0,0.0310,0.0362,0.000000,0.1170,0.555,134.948,82,1990



Last 5 rows:


,Track,Artist,Duration,Time_Signature,Danceability,Energy,Key,Loudness,Mode,Speechiness,Acousticness,Instrumentalness,Liveness,Valence,Tempo,Popularity,Year
914,I’m Your Angel,R. Kelly and Celine Dion,3:12,4,0.648,0.331,4,-10.952,1,0.0274,0.7760,0.000445,0.126,0.507,98.572,28,1999
915,Taking Everything,Gerald Levert,5:25,4,0.689,0.710,3,-5.395,0,0.0413,0.0886,0.000000,0.107,0.306,113.722,26,1999
916,What’s It Gonna Be?!,Busta Rhymes featuring Janet,2:41,4,0.710,0.742,9,-4.699,0,0.0339,0.1780,0.000064,0.080,0.504,127.969,26,1999
917,Almost Doesn’t Count,Brandy,3:38,4,0.682,0.543,0,-7.063,0,0.0454,0.4940,0.000000,0.158,0.478,71.582,19,1999
918,It Ain’t My Fault 1 & 2,Silkk the Shocker Featuring Mystikal,3:30,4,0.743,0.567,4,-12.468,0,0.2570,0.1960,0.000000,0.228,0.755,90.096,0,1999



Data Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 919 entries, 0 to 918
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Track             919 non-null    object 
 1   Artist            919 non-null    object 
 2   Duration          919 non-null    object 
 3   Time_Signature    919 non-null    int64  
 4   Danceability      919 non-null    float64
 5   Energy            919 non-null    float64
 6   Key               919 non-null    int64  
 7   Loudness          919 non-null    float64
 8   Mode              919 non-null    int64  
 9   Speechiness       919 non-null    float64
 10  Acousticness      919 non-null    float64
 11  Instrumentalness  919 non-null    float64
 12  Liveness          919 non-null    float64
 13  Valence           919 non-null    float64
 14  Tempo             919 non-null    float64
 15  Popularity        919 non-null    int64  
 16  Year              919 non-null 

#### 2. Variable Types


In [18]:
# Check data types
print("Data Types:")
print(df.dtypes)


Data Types:
Track                object
Artist               object
Duration             object
Time_Signature        int64
Danceability        float64
Energy              float64
Key                   int64
Loudness            float64
Mode                  int64
Speechiness         float64
Acousticness        float64
Instrumentalness    float64
Liveness            float64
Valence             float64
Tempo               float64
Popularity            int64
Year                  int64
dtype: object


#### 3. Missing Values


In [20]:
# Total missing values per column
missing_counts = df.isnull().sum()
missing_counts = missing_counts[missing_counts > 0]

print("Missing Values:")
print(missing_counts)
print("\nNo missing values detected.")

Missing Values:
Series([], dtype: int64)

No missing values detected.


There are **no missing values** in the dataset, so no imputation or data cleaning is needed for nulls.


### SECTION II: DATA CLEANING & INDEXING


---



#### 4. Check for and Remove Duplicates


In [53]:
# Checking for duplicates
duplicates = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

# Removing duplicates
df = df.drop_duplicates()

print(f"New shape after removing duplicates: {df.shape}")


Number of duplicate rows: 0
New shape after removing duplicates: (919, 19)


We found 0 duplicate rows in the dataset. These were removed using .drop_duplicates(), resulting in a new shape of 919 x 17.

In [55]:
# Convert to categorical
print(df['Time_Signature'].dtype)
print(df['Time_Signature'].unique())
df['Time_signature'] = df['Time_Signature'].astype('category')

int64
[4 3 1 5]


Converting the Time Signature values to categorical to make it a cleaner dataset.

#### 5. Filtering Using Logical Conditions

In [50]:
# Filtering songs that are likely vocal
vocal_songs = df[df['Instrumentalness'] < 0.2]

print(f"Number of vocal songs: {vocal_songs.shape[0]}")
vocal_songs.head()

Number of vocal songs: 850


,Track,Artist,Duration,Time_Signature,Danceability,Energy,Key,Loudness,Mode,Speechiness,Acousticness,Instrumentalness,Liveness,Valence,Tempo,Popularity,Year,Tempo_range,Time_signature
0,Everybody Everybody,Black Box,4:11,4,0.645,0.795,7,-12.095,1,0.0527,0.3470,0.003890,0.1040,0.535,112.067,87,1990,Fast (110-130 BPM),4
1,Close To You,Maxi Priest,3:45,4,0.590,0.727,6,-7.955,1,0.0706,0.0259,0.000065,0.0827,0.356,121.966,85,1990,Fast (110-130 BPM),4
2,Do Me!,Bell Biv DeVoe,3:56,4,0.615,0.534,9,-6.719,0,0.0386,0.1060,0.000018,0.0607,0.193,135.917,84,1990,Very Fast (130+ BPM),4
4,What It Takes,Aerosmith,3:21,4,0.672,0.655,10,-5.021,0,0.0310,0.0362,0.000000,0.1170,0.555,134.948,82,1990,Very Fast (130+ BPM),4
5,Alright,Janet Jackson,3:39,4,0.796,0.766,7,-5.974,1,0.2380,0.0742,0.000000,0.0827,0.558,110.034,80,1990,Fast (110-130 BPM),4


I filtered songs where Instrumentalness is less than 0.2 to capture tracks that are likely to have vocals.

In [51]:
# Filtering songs with high loudness
loud_songs = df[df['Loudness'] > -5]

print(f"Number of loud songs: {loud_songs.shape[0]}")
loud_songs.head()

Number of loud songs: 139


,Track,Artist,Duration,Time_Signature,Danceability,Energy,Key,Loudness,Mode,Speechiness,Acousticness,Instrumentalness,Liveness,Valence,Tempo,Popularity,Year,Tempo_range,Time_signature
14,Dangerous,Roxette,2:27,4,0.635,0.772,7,-4.825,1,0.0279,0.21600,0.000000,0.285,0.948,119.017,76,1990,Fast (110-130 BPM),4
16,Just A Friend,Biz Markie,4:01,4,0.822,0.739,1,-4.541,0,0.0311,0.02640,0.000043,0.181,0.745,118.990,75,1990,Fast (110-130 BPM),4
28,Ice Ice Baby,Vanilla Ice,4:14,4,0.979,0.792,2,-2.784,1,0.0388,0.00250,0.000008,0.102,0.583,116.012,70,1990,Fast (110-130 BPM),4
31,Pray,M.C. Hammer,3:19,4,0.200,0.977,3,-3.494,0,0.1590,0.00165,0.000011,0.259,0.390,195.711,68,1990,Very Fast (130+ BPM),4
32,Ready Or Not,After 7,3:20,4,0.715,0.872,2,-3.835,1,0.0509,0.00351,0.000038,0.107,0.748,93.043,68,1990,Moderate (90-110 BPM),4


I filtered for songs with Loudness greater than -5 dB, capturing tracks that are significantly louder than average.

### SECTION III: DESCRIPTIVE STATISTICS

#### 6. Summary Statistics for Numerical Columns

In [30]:
# Get descriptive statistics for numerical features
df.describe()

,Time_Signature,Danceability,Energy,Key,Loudness,Mode,Speechiness,Acousticness,Instrumentalness,Liveness,Valence,Tempo,Popularity,Year
count,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000,919.000000
mean,3.947769,0.622967,0.623548,5.412405,-8.349071,0.655060,0.075793,0.230068,0.052947,0.171865,0.531424,118.711789,57.722524,1994.405876
std,0.286769,0.154782,0.203975,3.514783,3.892821,0.475608,0.079570,0.256284,0.180214,0.138296,0.246091,29.684070,17.599522,2.926405
min,1.000000,0.135000,0.001340,0.000000,-42.950000,0.000000,0.023000,0.000004,0.000000,0.025000,0.006720,48.718000,0.000000,1990.000000
25%,4.000000,0.518500,0.479500,2.000000,-10.166500,0.000000,0.032400,0.029700,0.000000,0.089200,0.341500,96.924000,49.000000,1992.000000
50%,4.000000,0.631000,0.639000,6.000000,-7.599000,1.000000,0.042900,0.120000,0.000006,0.118000,0.535000,115.010000,60.000000,1994.000000
75%,4.000000,0.738000,0.780000,9.000000,-5.751000,1.000000,0.075600,0.344500,0.000741,0.217500,0.728500,133.424000,70.000000,1997.000000
max,5.000000,0.979000,0.999000,11.000000,0.919000,1.000000,0.529000,0.995000,0.987000,0.965000,0.975000,219.973000,92.000000,1999.000000


Interesting insight: The average Danceability score is around .623, which suggest that 90s hits were generally usually rhythmically engaging.



#### 7. Categorical Analysis

In [32]:
# Most common artists in the dataset
df['Artist'].value_counts().head(10)

,count
Artist,
Mariah Carey,16
Madonna,14
Janet Jackson,13
Boyz II Men,11
Whitney Houston,10
TLC,9
Michael Bolton,8
Bryan Adams,7
Aerosmith,7


The most common 3 artists were Mariah Carey, Madonna, and Jannet Jackson.

In [39]:
# Creating tempo bins
tempo_bins = [0, 90, 110, 130, 200]
tempo_labels = ['Slow (0-90 BPM)', 'Moderate (90-110 BPM)', 'Fast (110-130 BPM)', 'Very Fast (130+ BPM)']
df['Tempo_range'] = pd.cut(df['Tempo'], bins=tempo_bins, labels=tempo_labels)

# Finding Average danceability by tempo group
df.groupby('Tempo_range')['Danceability'].mean().sort_values(ascending=False)


<ipython-input-39-77f17ed52d5b>:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('Tempo_range')['Danceability'].mean().sort_values(ascending=False)


,Danceability
Tempo_range,
Moderate (90-110 BPM),0.678658
Fast (110-130 BPM),0.667020
Very Fast (130+ BPM),0.568765
Slow (0-90 BPM),0.553685


Songs in the Fast (110–130 BPM) category had the highest average Danceability, suggesting that upbeat tempos drive rhythm and movement.


#### 8. Grouped Summary Statistics

Key vs. Loudness

In [41]:
# Average loudness by key
df.groupby('Key')['Loudness'].mean().sort_values(ascending=False)

,Loudness
Key,
11,-7.757194
1,-7.977745
5,-8.033725
6,-8.186521
3,-8.219500
2,-8.365185
4,-8.427219
9,-8.438879
7,-8.458894


This tells us which musical keys tend to have louder more intense/energetic tracks.


In [58]:
# Count songs per time signature
df['Time_signature'].value_counts().sort_index()



,count
Time_signature,
1,3
3,45
4,865
5,6


Danceability Avg. vs. Time Signature

In [59]:
df.groupby('Time_Signature')['Danceability'].mean()

,Danceability
Time_Signature,
1,0.340333
3,0.474622
4,0.632779
5,0.462333


We converted the Time_Signature column to categorical format to better represent musical structure. Most songs are in '4/4', which reflects the standard rhythm of 90s music.


### SECTION IV: OBSERVATIONS AND QUESTIONS

#### Key Takeaways

- Most songs are in 4/4 time signature, confirming that mainstream 90s hits followed the standard rhythm structure common in pop, R&B, and hip-hop.

- The average danceability across the dataset is relatively high, suggesting that 90s hits were generally crafted to be rhythm-driven and engaging for listeners, possibly reflecting the popularity of club and radio-friendly music during the decade.


#### Further Points of Explorations :)

- Does danceability increase with tempo?

- Which musical key is most associated with high valence?
